pkg/ccl/backupccl/restore_data_processor.go

```diff
@@ -209,23 +222,88 @@ func inputReader(
 	}
 }
 
-func (rd *restoreDataProcessor) runRestoreWorkers(entries chan execinfrapb.RestoreSpanEntry) error {
+type mergedSST struct {
+	entry   execinfrapb.RestoreSpanEntry
+	iter    storage.SimpleMVCCIterator
+	cleanup func()
+}
+
+func (rd *restoreDataProcessor) openSSTs(
+	entry execinfrapb.RestoreSpanEntry, sstCh chan mergedSST,
+) error {
+	ctx := rd.Ctx
+
+	// The sstables only contain MVCC data and no intents, so using an MVCC
+	// iterator is sufficient.
+	var iters []storage.SimpleMVCCIterator
+	var dirs []cloud.ExternalStorage
+
+	// sendIters sends all of the currently accumulated iterators over the
+	// channel.
+	sendIters := func(itersToSend []storage.SimpleMVCCIterator, dirsToSend []cloud.ExternalStorage) {
+		iter := storage.MakeMultiIterator(itersToSend)
+
+		cleanup := func() {
+			for _, dir := range dirsToSend {
+				if err := dir.Close(); err != nil {
+					log.Warningf(ctx, "close export storage failed %v", err)
+				}
+			}
+			for _, iter := range itersToSend {
+				iter.Close()
+			}
+			iter.Close()
+		}
+
+		sstCh <- mergedSST{
+			entry:   entry,
+			iter:    iter,
+			cleanup: cleanup,
+		}
+
+		iters = make([]storage.SimpleMVCCIterator, 0)
+		dirs = make([]cloud.ExternalStorage, 0)
+	}
+
+	log.VEventf(rd.Ctx, 1 /* level */, "ingesting span [%s-%s)", entry.Span.Key, entry.Span.EndKey)
+
+	for _, file := range entry.Files {
+		log.VEventf(ctx, 2, "import file %s which starts at %s", file.Path, entry.Span.Key)
+
+		dir, err := rd.flowCtx.Cfg.ExternalStorage(ctx, file.Dir)
+		if err != nil {
+			return err
```

Comment:
if we bail early here, I think we have a bunch of unclosed dirs and iters?

Reply:
Augmented the processor's closing to drain any remaining ssts and cleanup after them.

Reply:
If we bail here though, before we call `sendIters` don't we have some dirs and iters in the local vars that nobody will close?

Reply:
Ah, I was looking at the wrong `return err` -- sorry for all the back and forth. Updated so that someone (either this method or the callback) always has responsibility for closing these. I also think we need to select on ctx.Done() to avoid blocking on this send forever, so I also added that.